src/Raven.Server/Rachis/Leader.cs

```diff
@@ -308,80 +308,90 @@ private void Run()
-                    // kick out of the cluster
-                    if (_previousPeersWereDisposed > 0) // Not Interlocked, because the race here is not interesting.
-                        continue;
+                        // don't truncate if we are disposing an old peer
+                        // otherwise he would not receive notification that he was
+                        // kick out of the cluster
+                        if (_previousPeersWereDisposed > 0) // Not Interlocked, because the race here is not interesting.
+                            continue;
 
-                    var lowestIndexInEntireCluster = GetLowestIndexInEntireCluster(out var lastTruncated);
-                    if (lowestIndexInEntireCluster == 0) // one of the nodes might be during the handshake
-                        continue;
+                        var lowestIndexInEntireCluster = GetLowestIndexInEntireCluster(out var lastTruncated);
+                        if (lowestIndexInEntireCluster == 0) // one of the nodes might be during the handshake
+                            continue;
 
-                    if (lowestIndexInEntireCluster > lastTruncated)
+                        if (lowestIndexInEntireCluster > lastTruncated)
+                        {
+                            using (_engine.ContextPool.AllocateOperationContext(out ClusterOperationContext context))
+                            using (context.OpenWriteTransaction())
+                            {
+                                _engine.TruncateLogBefore(context, lowestIndexInEntireCluster);
+                                LowestIndexInEntireCluster = lowestIndexInEntireCluster;
+                                context.Transaction.Commit();
+                            }
+                        }
+                    }
+                    catch (Exception ex)
                     {
+                        ClusterTopology clusterTopology;
                         using (_engine.ContextPool.AllocateOperationContext(out ClusterOperationContext context))
-                        using (context.OpenWriteTransaction())
+                        using (context.OpenReadTransaction())
+                        {
+                            clusterTopology = _engine.GetTopology(context);
+                        }
+
+                        if (clusterTopology.Members.Count == 1 && clusterTopology.Members.ContainsKey(_engine.Tag))
                         {
-                            _engine.TruncateLogBefore(context, lowestIndexInEntireCluster);
-                            LowestIndexInEntireCluster = lowestIndexInEntireCluster;
-                            context.Transaction.Commit();
+                            LogAndNotifyLeaderRunExceptions(ex);
+                            Task.Run(async () =>
+                            {
+                                await Task.Delay(100);
+                                _newEntry.Set();
+                            });
```

Comment:
Why are you doing this?

Reply:
Because once you failed to apply a command on "OnVoterConfirmation" and you caught here the exception (in single member cluster), In the next iteration of the loop you will be stuck in "switch (WaitHandle.WaitAny(handles, _engine.ElectionTimeout))".